apps/demo/src/routes/logcat.tsx

```diff
@@ -0,0 +1,230 @@
+import { IconButton, SearchBox, Stack, StackItem } from '@fluentui/react';
+import { AdbShell } from '@yume-chan/adb';
+import { encodeUtf8 } from '@yume-chan/adb-backend-webusb';
+import { AutoDisposable } from '@yume-chan/event';
+import { CSSProperties, useCallback, useContext, useEffect, useRef, useState } from 'react';
+import { Terminal } from 'xterm';
+import { FitAddon } from 'xterm-addon-fit';
+import { SearchAddon } from 'xterm-addon-search';
+import 'xterm/css/xterm.css';
+import { ErrorDialogContext } from '../components/error-dialog';
+import { ResizeObserver, withDisplayName } from '../utils';
+import { RouteProps } from './type';
+
+const ResizeObserverStyle: CSSProperties = {
+    width: '100%',
+    height: '100%',
+};
+
+const UpIconProps = { iconName: 'ChevronUp' };
+const DownIconProps = { iconName: 'ChevronDown' };
+
+class AdbTerminal extends AutoDisposable {
+    public terminal: Terminal = new Terminal({
+        scrollback: 9000,
+    });
+
+    public findAddon = new SearchAddon();
+
+    private readonly fitAddon = new FitAddon();
+
+    private _parent: HTMLElement | undefined;
+    public get parent() { return this._parent; }
+    public set parent(value) {
+        this._parent = value;
+
+        if (value) {
+            this.terminal.open(value);
+            this.fit();
+        }
+    }
+
+    private _shell: AdbShell | undefined;
+    public get socket() { return this._shell; }
+    public set socket(value) {
+        if (this._shell) {
+            this.dispose();
+        }
+
+        this._shell = value;
+
+        if (value) {
+            this.terminal.clear();
+            this.terminal.reset();
+
+            this.addDisposable(value.onStdout(data => {
+                this.terminal.write(new Uint8Array(data));
+            }));
+            this.addDisposable(value.onStderr(data => {
+                this.terminal.write(new Uint8Array(data));
+            }));
+            this.addDisposable(this.terminal.onData(data => {
+                const buffer = encodeUtf8(data);
+                value.write(buffer);
+            }));
+
+            this.fit();
+        }
+    }
+
+    public constructor() {
+        super();
+
+        this.terminal.loadAddon(this.findAddon);
+        this.terminal.loadAddon(this.fitAddon);
+    }
+
+    public fit() {
+        this.fitAddon.fit();
+        const { rows, cols } = this.terminal;
+        this._shell?.resize(rows, cols);
+    }
+}
+
+export interface LogcatRouteProps extends RouteProps {
+    applicationId: string | null;
+}
+
+export const Logcat = withDisplayName('Logcat')(({
+    visible,
+    device,
+    applicationId,
```

Comment:
What will happen if `appId` is `undefined`? Will it output all logs from the device?